oxd-server/src/main/java/org/gluu/oxd/server/op/Validator.java

```diff
@@ -142,14 +143,22 @@ public void validateAuthorizationCode(String code) {
     public static AbstractJwsSigner createJwsSigner(Jwt idToken, OpenIdConfigurationResponse discoveryResponse, PublicOpKeyService keyService, OpClientFactory opClientFactory, Rp rp) {
         final String algorithm = idToken.getHeader().getClaimAsString(JwtHeaderName.ALGORITHM);
         final SignatureAlgorithm signatureAlgorithm = SignatureAlgorithm.fromString(algorithm);
-        final String kid = idToken.getHeader().getClaimAsString(JwtHeaderName.KEY_ID);
         final String jwkUrl = discoveryResponse.getJwksUri();
+        String kid = idToken.getHeader().getClaimAsString(JwtHeaderName.KEY_ID);
 
         if (signatureAlgorithm == null)
             throw new HttpException(ErrorResponseCode.INVALID_ALGORITHM);
 
+        if (Strings.isNullOrEmpty(kid) && (signatureAlgorithm.getFamily() == AlgorithmFamily.RSA || signatureAlgorithm.getFamily() == AlgorithmFamily.EC)) {
+            LOG.warn("`kid` is missing in id_token. Checking keystore for `kid`");
+            kid = keyService.getKeyId(idToken, jwkUrl, signatureAlgorithm, Use.SIGNATURE);
+            if (Strings.isNullOrEmpty(kid)) {
+                throw new HttpException(ErrorResponseCode.KEY_ID_NOT_FOUND);
+            }
+        }
+
         if (signatureAlgorithm.getFamily() == AlgorithmFamily.RSA) {
-            final RSAPublicKey publicKey = keyService.getRSAPublicKey(jwkUrl, kid);
+            final RSAPublicKey publicKey = JwkClient.getRSAPublicKey(jwkUrl, kid);
```

Comment:
We should use `keyService` here, intention is to cache key temporary so under heavy load we don't download keys all the time.

Reply:
Done. We are caching the keys now.